map/autoware_lanelet2_divider/autoware_lanelet2_divider/xml_tool/xml_tool.py

```diff
@@ -0,0 +1,157 @@
+    )
+
+    whole_map_xml = ET.parse(input_osm_file_path)
+    whole_map_root = whole_map_xml.getroot()
+
+    node_list = []
+    way_list = []
+    relation_list = []
+
+    select_tags(whole_map_root, node_list, way_list, relation_list)
+
+    osm_files = [f for f in os.listdir(input_extracted_osm_folder) if f.endswith(".osm")]
+    for osm_file in tqdm(
+        osm_files,
+        desc=Debug.get_log("Completing missing elements in osm file", DebugMessageType.INFO),
+    ):
+        divided_map_xml = ET.parse(input_extracted_osm_folder + "/" + osm_file)
+        divided_map_root = divided_map_xml.getroot()
+
+        divided_node_list = []
+        divided_way_list = []
+        divided_relation_list = []
+
+        select_tags(divided_map_root, divided_node_list, divided_way_list, divided_relation_list)
+
+        for relation in divided_relation_list:
+            for r in relation.iter("member"):
+                if r.attrib["type"] == "way":
+                    if r.attrib["ref"] not in [way.attrib["id"] for way in divided_way_list]:
+                        for way in way_list:
+                            if way.attrib["id"] == r.attrib["ref"]:
+                                divided_map_root.append(way)
+                        select_tags(
+                            divided_map_root,
+                            divided_node_list,
+                            divided_way_list,
+                            divided_relation_list,
+                        )
+                elif r.attrib["type"] == "relation":
+                    if r.attrib["ref"] not in [rela.attrib["id"] for rela in divided_relation_list]:
+                        for rel in relation_list:
+                            if rel.attrib["id"] == r.attrib["ref"]:
+                                divided_map_root.append(rel)
+                        select_tags(
+                            divided_map_root,
+                            divided_node_list,
+                            divided_way_list,
+                            divided_relation_list,
+                        )
+
+        # Iterate on divided map's ways and find missing nodes
+        for way in divided_way_list:
+            for n in [nd.attrib["ref"] for nd in way.iter("nd")]:
+                if n not in [node.attrib["id"] for node in divided_node_list]:
+                    # find the node in the whole map and add it to the divided map
+                    for node in node_list:
+                        if node.attrib["id"] == n:
+                            divided_map_root.append(node)
+                    select_tags(
+                        divided_map_root,
+                        divided_node_list,
+                        divided_way_list,
+                        divided_relation_list,
+                    )
+
+        divided_map_xml.write(input_extracted_osm_folder + "/" + osm_file)
+
+    return True
+
+
+def complete_missing_version_tag(input_osm_file_path: str):
```

Comment:
What is the purpose of adding a version tag?

Reply:
Osmium throws an error similar to the following when there is missing version tag.

```
[ 0:00]   osmium version 1.16.0 (v1.16.0-4-g9c9f592)
[ 0:00]   libosmium version 2.18.0
[ 0:00] Command line options and default settings:
[ 0:00]   input options:
[ 0:00]     file name: /home/bzeren/Downloads/test/lanelet2_map_temp.osm
[ 0:00]     file format: 
[ 0:00]   output options:
[ 0:00]     file name: 
[ 0:00]     file format: 
[ 0:00]     generator: osmium/1.16.0
[ 0:00]     overwrite: yes
[ 0:00]     fsync: no
[ 0:00]   strategy options:
[ 0:00]     strategy: complete_ways
[ 0:00]     with history: no
[ 0:00]   other options:
[ 0:00]     config file: /home/bzeren/Downloads/test/output/config1.json
[ 0:00]     output directory: 
[ 0:00]     attributes to clean: (none)
[ 0:00] 
[ 0:00] Reading config file...
[ 0:00]   Directory set to '/home/bzeren/Downloads/test/output/lanelet2_map.osm'.
[ 0:00]   Reading extracts from config file...
[ 0:00]     Looking at extract '1.osm'...
[ 0:00]     Looking at extract '2.osm'...
[ 0:00]     Looking at extract '3.osm'...
[ 0:00]     Looking at extract '4.osm'...
[ 0:00] 
[ 0:00] Extracts:
[ 0:00] [01] Output:      /home/bzeren/Downloads/test/output/lanelet2_map.osm/1.osm
[ 0:00]      Format:      XML
[ 0:00]      Description: optional description
[ 0:00]      Envelope:    (28.8790289,41.0136999,28.8911729,41.0228977)
[ 0:00]      Type:        polygon
[ 0:00]      Geometry:    MULTIPOLYGON(((28.8790289 41.0138943,28.8909153 41.0136999,28.8911729 41.0227032,28.8792848 41.0228977,28.8790289 41.0138943)))
[ 0:00] [02] Output:      /home/bzeren/Downloads/test/output/lanelet2_map.osm/2.osm
[ 0:00]      Format:      XML
[ 0:00]      Description: optional description
[ 0:00]      Envelope:    (28.8909153,41.0135042,28.9030608,41.0227032)
[ 0:00]      Type:        polygon
[ 0:00]      Geometry:    MULTIPOLYGON(((28.8909153 41.0136999,28.9028016 41.0135042,28.9030608 41.0225074,28.8911729 41.0227032,28.8909153 41.0136999)))
[ 0:00] [03] Output:      /home/bzeren/Downloads/test/output/lanelet2_map.osm/3.osm
[ 0:00]      Format:      XML
[ 0:00]      Description: optional description
[ 0:00]      Envelope:    (28.8792848,41.0227032,28.8914306,41.031901)
[ 0:00]      Type:        polygon
[ 0:00]      Geometry:    MULTIPOLYGON(((28.8792848 41.0228977,28.8911729 41.0227032,28.8914306 41.0317064,28.8795409 41.031901,28.8792848 41.0228977)))
[ 0:00] [04] Output:      /home/bzeren/Downloads/test/output/lanelet2_map.osm/4.osm
[ 0:00]      Format:      XML
[ 0:00]      Description: optional description
[ 0:00]      Envelope:    (28.8911729,41.0225074,28.9033201,41.0317064)
[ 0:00]      Type:        polygon
[ 0:00]      Geometry:    MULTIPOLYGON(((28.8911729 41.0227032,28.9030608 41.0225074,28.9033201 41.0315106,28.8914306 41.0317064,28.8911729 41.0227032)))
[ 0:00] 
WARNING: Ignoring unknown option 'types' for 'complete_ways' strategy.
Can not read file without version (missing version attribute on osm element).
```